src/digest/md5.cr

```diff
@@ -1,254 +1,24 @@
-
-    # Round 3
-    a = hh(a, b, c, d, input[5], S31, 4294588738)  # 33
-    d = hh(d, a, b, c, input[8], S32, 2272392833)  # 34
-    c = hh(c, d, a, b, input[11], S33, 1839030562) # 35
-    b = hh(b, c, d, a, input[14], S34, 4259657740) # 36
-    a = hh(a, b, c, d, input[1], S31, 2763975236)  # 37
-    d = hh(d, a, b, c, input[4], S32, 1272893353)  # 38
-    c = hh(c, d, a, b, input[7], S33, 4139469664)  # 39
-    b = hh(b, c, d, a, input[10], S34, 3200236656) # 40
-    a = hh(a, b, c, d, input[13], S31, 681279174)  # 41
-    d = hh(d, a, b, c, input[0], S32, 3936430074)  # 42
-    c = hh(c, d, a, b, input[3], S33, 3572445317)  # 43
-    b = hh(b, c, d, a, input[6], S34, 76029189)    # 44
-    a = hh(a, b, c, d, input[9], S31, 3654602809)  # 45
-    d = hh(d, a, b, c, input[12], S32, 3873151461) # 46
-    c = hh(c, d, a, b, input[15], S33, 530742520)  # 47
-    b = hh(b, c, d, a, input[2], S34, 3299628645)  # 48
-
-    # Round 4
-    a = ii(a, b, c, d, input[0], S41, 4096336452)  # 49
-    d = ii(d, a, b, c, input[7], S42, 1126891415)  # 50
-    c = ii(c, d, a, b, input[14], S43, 2878612391) # 51
-    b = ii(b, c, d, a, input[5], S44, 4237533241)  # 52
-    a = ii(a, b, c, d, input[12], S41, 1700485571) # 53
-    d = ii(d, a, b, c, input[3], S42, 2399980690)  # 54
-    c = ii(c, d, a, b, input[10], S43, 4293915773) # 55
-    b = ii(b, c, d, a, input[1], S44, 2240044497)  # 56
-    a = ii(a, b, c, d, input[8], S41, 1873313359)  # 57
-    d = ii(d, a, b, c, input[15], S42, 4264355552) # 58
-    c = ii(c, d, a, b, input[6], S43, 2734768916)  # 59
-    b = ii(b, c, d, a, input[13], S44, 1309151649) # 60
-    a = ii(a, b, c, d, input[4], S41, 4149444226)  # 61
-    d = ii(d, a, b, c, input[11], S42, 3174756917) # 62
-    c = ii(c, d, a, b, input[2], S43, 718787259)   # 63
-    b = ii(b, c, d, a, input[9], S44, 3951481745)  # 64
-
-    @buf[0] &+= a
-    @buf[1] &+= b
-    @buf[2] &+= c
-    @buf[3] &+= d
-  end
-
-  private def final_impl(dst : Bytes) : Nil
-    tmp_in = uninitialized UInt32[16]
-
-    # save number of bits
-    tmp_in[14] = @i[0]
-    tmp_in[15] = @i[1]
-
-    # compute number of bytes mod 64
-    mdi = ((@i[0] >> 3) & 0x3F).to_i32
-
-    # pad out to 56 mod 64
-    pad_len = (mdi < 56) ? (56 - mdi) : (120 - mdi)
-    update PADDING.to_unsafe, pad_len
-
-    # append length in bits and transform
-    ii = 0
-    14.times do |i|
-      tmp_in[i] = (@in[ii + 3].to_u32 << 24) |
-                  (@in[ii + 2].to_u32 << 16) |
-                  (@in[ii + 1].to_u32 << 8) |
-                  (@in[ii])
-      ii += 4
-    end
-    transform tmp_in
-
-    # store buffer in digest
-    ii = 0
-    4.times do |i|
-      dst[ii] = (@buf[i] & 0xff).to_u8
-      dst[ii + 1] = ((@buf[i] >> 8) & 0xFF).to_u8
-      dst[ii + 2] = ((@buf[i] >> 16) & 0xFF).to_u8
-      dst[ii + 3] = ((@buf[i] >> 24) & 0xFF).to_u8
-      ii += 4
-    end
-  end
-
-  def digest_size : Int32
-    16
+  def dup
+    self.class.new(dup_ctx)
   end
```

Comment:
Couldn't this be defined directly within the `Digest` class itself?

Reply:
No, becuase the `OpenSSL::Digest` has no constructor without algorithm name. And having another intermediate class sounds overkill to me.